Titan image genai commons (#8005)
mendix/docs: content/en/docs/appstore/use-content/modules/aws/amazon-bedrock.md

| --- | --- |
| `KnowledgeBaseId` | The KnowledgeBaseId attribute describes the unique identifier of the knowledge base that is queried and the foundation model used for generation. |

##### 4.1.1.8 TitanImageOptions_Extension {#titan-image-options-extension}

This entity extends the GenAICommons.ImageOptions entity with values specific to Titan Image models. To add an instance of this entity to an GenAICommons.ImageOptions object, the exposed [Image Generation: Add Titan Image Extension](#add-titan-image-extension) operation can be used.

| Attribute | Description |
| --- | --- |

> | `NegativeText` | The NegativeText attribute describes a text prompt to define what not to include in the image. |

Reviewer: Does it describe the text prompt? I think it **is** the text prompt?
```suggestion
| `NegativeText` | The NegativeText attribute is a string containing a prompt describing what not to include in the image. |
```

Author: I agree. Suggestion is fine

##### 4.1.1.9 TitanImageResponse {#titan-image-response}

This entity extends the GenAICommons.Response entity with values specific to Titan Image models.

| Attribute | Description |
| --- | --- |

> | `Error` | The Error attribute describes the error message if the request violates the content moderation policy. |

Reviewer: ```suggestion
| `Error` | The Error attribute contains the error message if the request violates the content moderation policy. |
```

Author: also fine. You can even shorten it: The Error attribute contains the error message if the request violates the content moderation policy.

##### 4.1.1.10 TitanImageMappingHelper {#titan-image-mapping-helper}

> This helper entity solely serves the purpose of creating a flat entity to produce the correct export mapping.

Reviewer: What do you mean by a "flat entity"?

Author: This is quite technical language which means that a data structure where several entities are connected over associations has been condensed into one entity instead.

Reviewer: How about:
```suggestion
This helper entity summarizes data from several associated entities into a single entity where it can be formatted to produce the correct export mapping.
```

Reviewer: @strategicalliances - still one more change for you to approve.

The history can be enabled using the `SessionId` parameter on the [RetrieveAndGenerateRequest_Extension](#retrieve-and-generate-request-extension) entity. By reusing the same `SessionId` value, the model will run in the context of the session. 

##### 4.2.1.4 Image Generation {#image-generation}

The `Image Generation` operation can be used to generate one or more images. This operation corresponds to the *ImageGeneration_AmazonBedrock* microflow. Currently 'Amazon Titan Image Generator G1' is the only supported model for image generation of the Amazon Bedrock Connector. 

The input and output for this service are shown in the table below:

| Input | Output |
| --- | --- |
| `UserPrompt (String)`,<br/>`AmazonBedrockConnection (object)`,<br/>`GenAICommons.ImageOptions (object, optional)`| `GenAICommons.Response (object)`|

> `GenAICommons.ImageOptions` is an optional parameter that can be omitted by passing *empty* instead. If provided, it allows to set additional options for Image Generation. 

Reviewer: If it has to be *empty* then it isn't optional, you have to pass something. How about:
```suggestion
| `UserPrompt (String)`,<br/>`AmazonBedrockConnection (object)`,<br/>`GenAICommons.ImageOptions (object)`| `GenAICommons.Response (object)`|

`GenAICommons.ImageOptions` can be an empty object. If provided, it allows you to set additional options for Image Generation. 
```

Author: fine. Or 'is optional'?

Reviewer: My point is that it isn't "optional". If you use this operation you have to provide something in `GenAICommons.ImageOptions` even if what you provide is an empty entity.

| --- | --- |
| `GenAICommons.Request (object)`, `KmsKeyArn (string, optional)`, `SessionId (string, optional)`, `Enum_RetrieveAndGenerateType (enumeration, optional)` | `RetrieveAndGenerateRequest_Extension (object)` |

##### 4.2.2.5 Image Generation: Add Titan Image Extension {#add-titan-image-extension}

Use this microflow to add a new [TitanImageOptions_Extension](#titan-image-options-extension) object to your GenAICommons.ImageOptions object. This will allow you to configure the *NegativeText* attribute.

This operation corresponds to the **TitanImageOptions_Extension_Create** microflow.

| Input | Output |
| --- | --- |

> | `GenAICommons.ImageOptions (object)`, `NegativeText (string, optional)` | `TitanImageOptions_Extension (object)` |

Reviewer: I think that you always have to supply a string here, so it shouldn't be optional (although it could be zero-length or spaces).
Also, there is no point in calling this microflow if you give it an empty NegativeText string as the only thing it does is add a NegativeText string to your image options.

Author: HI Mark, good find. Could you remove 'optional'?

Reviewer: Done

##### 4.2.2.6 Image Generation: Set Image Size (Titan Image) {#set-titan-image-size}

Use this microflow to set the *Height* and *Width* attributes of your GenAICommons.ImageOptions object to any valid image size supported by Titan Image models. The `ENUM_ImageSize_TitanImage` enumeration contains all valid height-width combinations to choose from.

This operation corresponds to the **ImageOptions_SetImageSize_TitanImage** microflow.

| Input | Output |
| --- | --- |
| `GenAICommons.ImageOptions (object)`, `ENUM_ImageSize_TitanImage (enumeration)` | `none` |

##### 4.2.2.7 Image Generation: Set Randomness {#set-randomness}

Use this microflow to set the *Seed* and *CfgScale* attributes of your GenAICommons.ImageOptions object. These attributes can be used to influence the randomness of the image generation.

For more information, please refer to the specific model documentation such as [Titan Image Generator G1](https://docs.aws.amazon.com/bedrock/latest/userguide/model-parameters-titan-image.html).

This operation corresponds to the **ImageOptions_SetRandomness** microflow.

| Input | Output |
| --- | --- |

> | `GenAICommons.ImageOptions (object)`, `Seed (integer, optional)`, `CfgScale (decimal, optional)` | `none` |

Reviewer: I don't think the Seed and CfgScale are optional, I think you will have to provide something as Microflows do not have optional parameters? Which values will be replaced by defaults? zero?

Author: By optional we mean that it is fine to set them 'empty' and then they are not part of the request at all. They are just not sent to the API.

Reviewer: Added this as an explanation.
